# Task 2: Data Cleaning & Preprocessing

## Dataset: Titanic Dataset

### Objectives
- Handle missing values
- Remove duplicate records
- Verify data types
- Detect and handle outliers
- Normalize numerical features
- Export cleaned dataset

# Import Libraries


In [5]:
import pandas as pd

In [1]:

import numpy as np

# Load Dataset

In [7]:
df = pd.read_csv("train.csv")

# View Dataset

In [ ]:
df.head()


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


# Dataset Shape

In [8]:

print("Shape:", df.shape)

Shape: (891, 12)


# Dataset Information

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


# Statistical Summary

In [ ]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


# Missing Value Analysis

In [ ]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


# Missing Value Handling



In [10]:
df["Age"] = df["Age"].fillna(df["Age"].median())

In [11]:
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

In [12]:
df.drop("Cabin", axis=1, inplace=True)

In [13]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


### Observation

- Age missing values filled using Median.
- Embarked missing values filled using Mode.
- Cabin column dropped due to excessive missing values.
- Dataset now contains no missing values.

# Duplicate Record Check

In [21]:
duplicate_count = df.duplicated().sum()

print("Number of Duplicate Records:", duplicate_count)

Number of Duplicate Records: 0


In [23]:
df = df.drop_duplicates()

print("Duplicates After Removal:", df.duplicated().sum())

Duplicates After Removal: 0


# Data Type Verification

In [17]:
df.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


# Outlier Detection

In [18]:
Q1 = df["Fare"].quantile(0.25)

Q3 = df["Fare"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR

upper_bound = Q3 + 1.5 * IQR

outliers = df[
    (df["Fare"] < lower_bound) |
    (df["Fare"] > upper_bound)
]

print("Number of Outliers:", len(outliers))

Number of Outliers: 116


In [19]:
df = df[
    (df["Fare"] >= lower_bound) &
    (df["Fare"] <= upper_bound)
]

In [24]:
print("Shape Before Outlier Removal:", df.shape)

Shape Before Outlier Removal: (775, 11)


### Observation

- IQR method was used for outlier detection.
- 116 Fare outliers were identified.
- Dataset size reduced from 891 rows to 775 rows after outlier treatment.

# Feature Scaling


In [25]:
from sklearn.preprocessing import MinMaxScaler

In [26]:
scaler = MinMaxScaler()

columns_to_scale = [
    "Age",
    "Fare",
    "SibSp",
    "Parch"
]

df[columns_to_scale] = scaler.fit_transform(
    df[columns_to_scale]
)

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,0.271174,0.2,0.0,A/5 21171,0.111538,S
2,3,1,3,"Heikkinen, Miss. Laina",female,0.321438,0.0,0.0,STON/O2. 3101282,0.121923,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,0.434531,0.2,0.0,113803,0.816923,S
4,5,0,3,"Allen, Mr. William Henry",male,0.434531,0.0,0.0,373450,0.123846,S
5,6,0,3,"Moran, Mr. James",male,0.346569,0.0,0.0,330877,0.130128,Q


In [27]:
df[["Age","Fare","SibSp","Parch"]].describe()

,Age,Fare,SibSp,Parch
count,775.000000,775.000000,775.000000,775.000000
mean,0.355978,0.274186,0.087484,0.056774
std,0.160620,0.208894,0.179968,0.130986
min,0.000000,0.000000,0.000000,0.000000
25%,0.271174,0.121474,0.000000,0.000000
50%,0.346569,0.200000,0.000000,0.000000
75%,0.421965,0.400000,0.200000,0.000000
max,1.000000,1.000000,1.000000,1.000000


# Export Clean Dataset

In [28]:
df.to_csv("clean_titanic.csv", index=False)

print("Clean dataset exported successfully.")

Clean dataset exported successfully.


# Final Dataset Verification

In [29]:
print("Final Shape:", df.shape)

print("\nMissing Values:")
print(df.isnull().sum())

df.head()

Final Shape: (775, 11)

Missing Values:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,0.271174,0.2,0.0,A/5 21171,0.111538,S
2,3,1,3,"Heikkinen, Miss. Laina",female,0.321438,0.0,0.0,STON/O2. 3101282,0.121923,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,0.434531,0.2,0.0,113803,0.816923,S
4,5,0,3,"Allen, Mr. William Henry",male,0.434531,0.0,0.0,373450,0.123846,S
5,6,0,3,"Moran, Mr. James",male,0.346569,0.0,0.0,330877,0.130128,Q


# Conclusion

The Titanic dataset was successfully cleaned and preprocessed.

Tasks Completed:
- Missing values handled
- Cabin column removed
- Duplicate records checked
- Data types verified
- Outliers detected and treated using the IQR method
- Numerical features normalized using MinMaxScaler
- Clean dataset exported successfully

The final dataset is ready for exploratory data analysis and machine learning applications.

In [30]:
print("Final Shape:", df.shape)

Final Shape: (775, 11)


In [31]:
print("Final Shape:", df.shape)

print("\nMissing Values:")
print(df.isnull().sum())

print("\nData Types:")
print(df.dtypes)

Final Shape: (775, 11)

Missing Values:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

Data Types:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp          float64
Parch          float64
Ticket          object
Fare           float64
Embarked        object
dtype: object
